primitives/keystore/src/lib.rs

```diff
@@ -45,109 +45,174 @@ pub enum Error {
 	///
-	/// Returns a set of public keys the signer supports.
-	fn keys(&self, id: KeyTypeId) -> Result<Vec<CryptoTypePublicPair>, Error>;
+	/// Returns a set of public keys the signer supports in raw format.
+	fn keys(&self, key_type: KeyTypeId) -> Result<Vec<Vec<u8>>, Error>;
 
 	/// Checks if the private keys for the given public key and key type combinations exist.
 	///
 	/// Returns `true` iff all private keys could be found.
 	fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> bool;
 
-	/// Sign with key
+	/// Convenience method to sign a message using the given key type and a raw public key
+	/// for secret lookup.
 	///
-	/// Signs a message with the private key that matches
-	/// the public key passed.
+	/// The message is signed using the cryptographic primitive specified by `crypto_id`.
+	///
+	/// Schemes supported by the default trait implementation: sr25519, ed25519 and ecdsa.
+	/// To support more schemes you can overwrite this method.
 	///
 	/// Returns the SCALE encoded signature if key is found and supported, `None` if the key doesn't
 	/// exist or an error when something failed.
 	fn sign_with(
 		&self,
 		id: KeyTypeId,
-		key: &CryptoTypePublicPair,
+		crypto_id: CryptoTypeId,
+		public: &[u8],
 		msg: &[u8],
-	) -> Result<Option<Vec<u8>>, Error>;
-
-	/// Generate VRF signature for given transcript data.
-	///
-	/// Receives KeyTypeId and Public key to be able to map
-	/// them to a private key that exists in the keystore which
-	/// is, in turn, used for signing the provided transcript.
-	///
-	/// Returns a result containing the signature data.
-	/// Namely, VRFOutput and VRFProof which are returned
-	/// inside the `VRFSignature` container struct.
-	///
-	/// This function will return `None` if the given `key_type` and `public` combination
-	/// doesn't exist in the keystore or an `Err` when something failed.
-	fn sr25519_vrf_sign(
-		&self,
-		key_type: KeyTypeId,
-		public: &sr25519::Public,
-		transcript_data: VRFTranscriptData,
-	) -> Result<Option<VRFSignature>, Error>;
-
-	/// Generate an ECDSA signature for a given pre-hashed message.
-	///
-	/// Receives [`KeyTypeId`] and an [`ecdsa::Public`] key to be able to map
-	/// them to a private key that exists in the keystore. This private key is,
-	/// in turn, used for signing the provided pre-hashed message.
-	///
-	/// The `msg` argument provided should be a hashed message for which an
-	/// ECDSA signature should be generated.
-	///
-	/// Returns an [`ecdsa::Signature`] or `None` in case the given `id` and
-	/// `public` combination doesn't exist in the keystore. An `Err` will be
-	/// returned if generating the signature itself failed.
-	fn ecdsa_sign_prehashed(
-		&self,
-		id: KeyTypeId,
-		public: &ecdsa::Public,
-		msg: &[u8; 32],
-	) -> Result<Option<ecdsa::Signature>, Error>;
+	) -> Result<Option<Vec<u8>>, Error> {
+		use codec::Encode;
+
+		let signature = match crypto_id {
+			sr25519::CRYPTO_ID => {
```

Comment:
I thought you wanted to remove `CRYPTO_ID` ?

Reply:
@bkchr 

Yes, that was indeed the awesome original idea.

Unfortunately, though, considering that AURA is generic over the used crypto scheme, we need a way to select the correct scheme to use.

The ideal solution (IIRC something like what you suggested during our chat), is to have a sign-with method generic over an object implementing `Pair`.
That is, something like [this](https://github.com/paritytech/substrate/blob/ea7238237da06731ff2c4280d71f00c0b361e74c/client/keystore/src/local.rs#L60-L65)

Thus a generic `sign_with` method would look like:

    fn sign_with<P: Pair>(&self, id: AppCryptoId, public: P::Public, msg: &[u8]) -> P::Signature;

**Sadly** a trait with a method like this can't be made into a trait-objejct (as it depends on a generic type) and we use the keystore as a `Arc<dyn Keystore>` all over the project.

Reply:
> **Sadly** a trait with a method like this can't be made into a trait-objejct (as it depends on a generic type) and we use the keystore as a `Arc<dyn Keystore>` all over the project.

Well, if you'd go through `&dyn` then you could, e.g.:

```rust
use std::sync::Arc;

trait Public {}
trait Pair {
    type Public: Public;
    type Signature: 'static;
}

trait Keystore {
    fn yo_sign_this_plz_impl(&self, pair: &dyn Public) -> Box<dyn std::any::Any>;
}

impl dyn Keystore {
    fn yo_sign_this_plz<P: Pair>(&self, pair: &P::Public) -> P::Signature {
        match self.yo_sign_this_plz_impl(pair).downcast() {
            Ok(sig) => *sig,
            Err(_) => unreachable!()
        }
    }
}

fn it_works<P: Pair>(keystore: Arc<dyn Keystore>, pubkey: P::Public) -> P::Signature {
    keystore.yo_sign_this_plz::<P>(&pubkey)
}
```

But (in this example) this would require that `Public` is now object safe, so you'd have to use similar tricks for it, and so on until you make everything object safe all the way to the bottom. It's turtles all the way down. Probably not worth it though?